cid.go

```diff
@@ -522,6 +525,74 @@ func (c Cid) Prefix() Prefix {
 	}
 }
 
+func (c Cid) MarshalCBOR(w io.Writer) error {
+	tag := cbg.CborEncodeMajorType(6, 42)
+	if _, err := w.Write(tag); err != nil {
+		return err
+	}
+
+	typ := cbg.CborEncodeMajorType(2, uint64(len(c.Bytes())+1))
+	if _, err := w.Write(typ); err != nil {
+		return err
+	}
+
+	// that binary multibase prefix...
+	if _, err := w.Write([]byte{0}); err != nil {
+		return err
+	}
+
+	if _, err := w.Write(c.Bytes()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *Cid) UnmarshalCBOR(br cbg.ByteReader) error {
+	maj, low, err := cbg.CborReadHeader(br)
+	if err != nil {
+		return err
+	}
+
+	if maj != 6 || low != 42 {
+		return fmt.Errorf("CBOR serialized CIDs must have the tag 42")
+	}
+
+	maj, low, err = cbg.CborReadHeader(br)
+	if err != nil {
+		return err
+	}
+
+	if maj != 2 {
+		return fmt.Errorf("CBOR serialized CIDs must be tagged byte arrays")
+	}
+
+	if low > 256 {
+		return fmt.Errorf("CIDs cannot be longer than 256 bytes")
+	}
+	if low < 2 {
+		return fmt.Errorf("encoded CID body must be at least two bytes long")
+	}
+
+	buf := make([]byte, low)
```

Comment:
Could we... stack allocate? Although go may catch this for us (or we could use a pool). Food for thought.

Reply:
I don't think we're able to force it here if go's escape analyzer doesnt like it